lib/compress.js

```diff
@@ -1117,10 +1134,11 @@ merge(Compressor.prototype, {
             throw def;
         });
         def(AST_Dot, function(compressor){
-            if (compressor.option("unsafe") && this.property == "length") {
-                var str = ev(this.expression, compressor);
-                if (typeof str == "string")
-                    return str.length;
+            if (compressor.option("unsafe")) {
+                var val = ev(this.expression, compressor);
+                if (HOP(val, this.property)) {
```

Comment:
Before calling `HOP()` must check if first argument is an object literal:
```
$ echo 'a = null.p;' | bin/uglifyjs -c unsafe
undefined:6244
                if (ex !== def) throw ex;
                                ^

TypeError: Cannot convert undefined or null to object
    at hasOwnProperty (<anonymous>)
```
We want to preserve bad javascript code as is. Garbage in, garbage out.